---
title: "Homework Assigment #1"
author: Jeffrey Lewis & Carson Stacy
date: "`r Sys.Date()`"
format: 
  html: 
    code-fold: show
  pdf: default
  docx: default
---

```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache=FALSE)
```

```{r}
#| message: false
#| warning: false
library(readr)
library(dplyr)
library(ggplot2)
library(cowplot)

set_theme(theme_minimal())
```


## Assignment #1: Due January 29th

**Please upload your completed assignments into Blackboard as a .qmd file so that we can reproduce your code when relevant. Save your file as StudentID_Assignment1.**

**General Rules:**

1\) We strongly encourage people to work in groups and take advantage of any resources at your disposal (including AI\--see the syllabus for our complete policy). However, we expect each student to demonstrate their own understanding by generating their own answers and not simply copying (which is plagiarism).

2\) Science is collaborative, and we encourage students to work through these problems in groups. Please do cite your collaborators if any.

3\) If you used any outside references, please cite them. Additionally if you used generative AI for the assignment, please provide a link to the prompt and answers. This will help us troubleshoot or provide feedback if the AI makes any mistakes.

4\) Sonali and I are happy to field questions about the assignments either in class, via email, or by appointment.

**Optional DataCamp Assignments:**

The class has a free 6-month subscription to DataCamp, which has a whole bunch of interactive courses that span beginner level through advanced. You are free to take any courses of interest for free during the semester, and I would like to point out that there is a sequence of courses for [analyzing genomic data](https://app.datacamp.com/learn/skill-tracks/analyzing-genomic-data-in-r) that students may like to use throughout the semester to complement the activities in class.

For this assignment period, I would highly recommend that students complete ["Introduction to R Chapter 1: Intro to basics"](https://app.datacamp.com/learn/courses/free-introduction-to-r) and ["Reporting with R Markdown Chapter 1: Getting Started with R Markdown."](https://app.datacamp.com/learn/courses/reporting-with-rmarkdown)

**The Main Assignment:**

Using the "01_GettingStartedInR.qmd" in class exercise, answer the following questions about the DE_results_msn24_EtOH dataset. As a reminder, this dataset includes the log~2~ fold changes of gene expression in an experiment examining the ethanol stress response for the YPS606 strain of *S. cerevisiae* and an *msn2/4ΔΔ* (transcription factor) mutant. Log~2~ fold changes for the ethanol response are measured by comparing gene expression in ethanol vs. an unstressed control (so positive values mean up-regulated by ethanol stress, and negative values mean down-regulated by ethanol). For WT vs mutant comparisons, negative values mean higher expression in the WT strain and positive values mean higher expression in the mutant.

```{r}
DE_results_msn24_EtOH = read_tsv("https://raw.githubusercontent.com/clstacy/GenomicAnalysis/main/Data/Msn24_EtOH/DE_genes_Msn24_EtOH.txt")
```


Answer the following questions (type your answers below each question, blank code chunks are included that you can paste or type commands into):

1.  How many genes are included in this study?

```{r}
length(unique(DE_results_msn24_EtOH$gene_id))
```

There are 5756 unique genes in the dataset.

2.  Which gene has the highest log fold change in the *msn2/4*ΔΔ mutant EtOH response?

```{r}
#| paged-print: true
# Reference the dataset
DE_results_msn24_EtOH %>%
  # Sort the dataset by the log fold-change of the mutant ethanol response (descending = highest first)
  arrange(desc(logFC_YPS606_msn24dd_EtOH_response)) %>%
  
  # Get the first value in the dataset, which will have the highest value of the mutant ethanol response
  head(1) 
```

The TDH1 gene, which codes for glyceraldehyde-3-phosphate dehydrogenase isozyme, has the highest log fold change in the *msn2/4*ΔΔ mutant EtOH response- ~9.06.

3.  How many heat shock protein (HSP) genes are differentially expressed (FDR \< 0.01) in unstressed conditions for the mutant?

```{r}
#| paged-print: true
# Reference the dataset
DE_results_msn24_EtOH %>%
  filter(
    # Get all rows where the FDR of the mutant unstressed condition is less than 0.01
    fdr_WT.vs.msn24dd_unstressed < 0.01,
    
    # Get all rows where the common name includes the string "HSP"
    grepl("HSP", common_name, ignore.case = TRUE)
  )
```

There are two HSP proteins that are differentially expressed (FDR \< 0.01) in unstressed conditions for the mutant.

4.  Do the genes with the largest magnitude fold changes [tend]{.underline} to have the smallest *p*-values? Why might there be cases where there's a mismatch between the magnitude of fold change and *p*-values?

```{r}

a1 <- ggplot(
  data = DE_results_msn24_EtOH, aes(x = logFC_YPS606_msn24dd_EtOH_response,
                                    y = -log10(pvalue_YPS606_msn24dd_EtOH_response))) +
  geom_point(alpha = 0.5) +
  labs(title = "msn2/4∆∆ EtOH response", 
       x = "logFC", 
       y = "-log(p value)")
a2 <- ggplot(
  data = DE_results_msn24_EtOH, aes(x = abs(logFC_YPS606_msn24dd_EtOH_response),
                                    y = -log10(pvalue_YPS606_msn24dd_EtOH_response))) +
  geom_point(alpha = 0.5) +
  geom_smooth(method = "lm") +
  labs(title = "Absolute msn2/4∆∆ EtOH response", 
       x = "|logFC|", 
       y = "-log(p value)")

b1 <- ggplot(
  data = DE_results_msn24_EtOH, aes(x = logFC_YPS606_WT_etoh_response,
                                    y = -log10(pvalue_YPS606_WT_etoh_response))) +
  geom_point(alpha = 0.5) +
  labs(title = "WT EtOH response", 
       x = "logFC", 
       y = "-log(p value)")
b2 <- ggplot(
  data = DE_results_msn24_EtOH, aes(x = abs(logFC_YPS606_WT_etoh_response),
                                    y = -log10(pvalue_YPS606_WT_etoh_response))) +
  geom_point(alpha = 0.5) +
  geom_smooth(method = "lm") +
  labs(title = "Absolute WT EtOH response", 
       x = "|logFC|", 
       y = "-log(p value)")

c1 <- ggplot(
  data = DE_results_msn24_EtOH, aes(x = logFC_WT.vs.msn24dd_EtOH_response,
                                    y = -log10(pvalue_WT.vs.msn24dd_EtOH_response))) +
  geom_point(alpha = 0.5) +
  labs(title = "WT vs msn2/4∆∆ EtOH response", 
       x = "logFC", 
       y = "-log(p value)")
c2 <- ggplot(
  data = DE_results_msn24_EtOH, aes(x = abs(logFC_WT.vs.msn24dd_EtOH_response),
                                    y = -log10(pvalue_WT.vs.msn24dd_EtOH_response))) +
  geom_point(alpha = 0.5) +
  geom_smooth(method = "lm") +
  labs(title = "Absolute WT vs msn2/4∆∆ EtOH response", 
       x = "|logFC|", 
       y = "-log(p value)")

plot_grid(b1, b2, a1, a2, c1, c2, ncol = 2, nrow = 3)
```

The genes with the largest magnitude fold change tend to have the smallest p-values (small p-value -> large -log(p-value) because large changes in expression are typically significant. There might be a mismatch between the magnitude of fold change and the p-values

5.  Which isoform of phosphoglucomutase is upregulated in response to ethanol stress, *PGM1* or *PGM2*? Do you think *msn2/4* is responsible for this difference? Why or why not?

```{r}
# Reference the dataset
DE_results_msn24_EtOH %>%
  # Filter
  filter(grepl("phosphoglucomutase", annotation, ignore.case = TRUE)) %>%
  arrange(common_name) %>%
  select(common_name, 
         logFC_YPS606_WT_etoh_response, 
         logFC_YPS606_msn24dd_EtOH_response, 
         logFC_WT.vs.msn24dd_EtOH_response)
```

**Collaborators:**

**References:**

**AI Use:**

System information for reproducibility:

```{r}
pander::pander(sessionInfo())
```
