<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jeffrey Lewis &amp; Carson Stacy">
<meta name="dcterms.date" content="2026-01-22">

<title>Homework Assigment #1</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="Homework_Assignment_1_files/libs/clipboard/clipboard.min.js"></script>
<script src="Homework_Assignment_1_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="Homework_Assignment_1_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="Homework_Assignment_1_files/libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="Homework_Assignment_1_files/libs/quarto-html/popper.min.js"></script>
<script src="Homework_Assignment_1_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Homework_Assignment_1_files/libs/quarto-html/anchor.min.js"></script>
<link href="Homework_Assignment_1_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Homework_Assignment_1_files/libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Homework_Assignment_1_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Homework_Assignment_1_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Homework_Assignment_1_files/libs/bootstrap/bootstrap-d6a003b94517c951b2d65075d42fb01b.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body class="quarto-light">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="Homework_Assignment_1.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li><li><a href="Homework_Assignment_1.docx"><i class="bi bi-file-word"></i>MS Word</a></li></ul></div></div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Homework Assigment #1</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Jeffrey Lewis &amp; Carson Stacy </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 22, 2026</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">set_theme</span>(<span class="fu">theme_minimal</span>())</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>
<section id="assignment-1-due-january-29th" class="level2">
<h2 class="anchored" data-anchor-id="assignment-1-due-january-29th">Assignment #1: Due January 29th</h2>
<p><strong>Please upload your completed assignments into Blackboard as a .qmd file so that we can reproduce your code when relevant. Save your file as StudentID_Assignment1.</strong></p>
<p><strong>General Rules:</strong></p>
<p>1) We strongly encourage people to work in groups and take advantage of any resources at your disposal (including AI--see the syllabus for our complete policy). However, we expect each student to demonstrate their own understanding by generating their own answers and not simply copying (which is plagiarism).</p>
<p>2) Science is collaborative, and we encourage students to work through these problems in groups. Please do cite your collaborators if any.</p>
<p>3) If you used any outside references, please cite them. Additionally if you used generative AI for the assignment, please provide a link to the prompt and answers. This will help us troubleshoot or provide feedback if the AI makes any mistakes.</p>
<p>4) Sonali and I are happy to field questions about the assignments either in class, via email, or by appointment.</p>
<p><strong>Optional DataCamp Assignments:</strong></p>
<p>The class has a free 6-month subscription to DataCamp, which has a whole bunch of interactive courses that span beginner level through advanced. You are free to take any courses of interest for free during the semester, and I would like to point out that there is a sequence of courses for <a href="https://app.datacamp.com/learn/skill-tracks/analyzing-genomic-data-in-r">analyzing genomic data</a> that students may like to use throughout the semester to complement the activities in class.</p>
<p>For this assignment period, I would highly recommend that students complete <a href="https://app.datacamp.com/learn/courses/free-introduction-to-r">“Introduction to R Chapter 1: Intro to basics”</a> and <a href="https://app.datacamp.com/learn/courses/reporting-with-rmarkdown">“Reporting with R Markdown Chapter 1: Getting Started with R Markdown.”</a></p>
<p><strong>The Main Assignment:</strong></p>
<p>Using the “01_GettingStartedInR.qmd” in class exercise, answer the following questions about the DE_results_msn24_EtOH dataset. As a reminder, this dataset includes the log<sub>2</sub> fold changes of gene expression in an experiment examining the ethanol stress response for the YPS606 strain of <em>S. cerevisiae</em> and an <em>msn2/4ΔΔ</em> (transcription factor) mutant. Log<sub>2</sub> fold changes for the ethanol response are measured by comparing gene expression in ethanol vs.&nbsp;an unstressed control (so positive values mean up-regulated by ethanol stress, and negative values mean down-regulated by ethanol). For WT vs mutant comparisons, negative values mean higher expression in the WT strain and positive values mean higher expression in the mutant.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>DE_results_msn24_EtOH <span class="ot">=</span> <span class="fu">read_tsv</span>(<span class="st">"https://raw.githubusercontent.com/clstacy/GenomicAnalysis/main/Data/Msn24_EtOH/DE_genes_Msn24_EtOH.txt"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 5756 Columns: 18
── Column specification ────────────────────────────────────────────────────────
Delimiter: "\t"
chr  (3): gene_id, common_name, annotation
dbl (15): logFC_YPS606_WT_etoh_response, pvalue_YPS606_WT_etoh_response, fdr...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(DE_results_msn24_EtOH)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 5,756
Columns: 18
$ gene_id                             &lt;chr&gt; "YMR105C", "YML100W", "YER053C", "…
$ common_name                         &lt;chr&gt; "PGM2", "TSL1", "PIC2", "NCE102", …
$ annotation                          &lt;chr&gt; "Phosphoglucomutase", "Large subun…
$ logFC_YPS606_WT_etoh_response       &lt;dbl&gt; 7.5999973, 7.7618280, 6.6940024, 0…
$ pvalue_YPS606_WT_etoh_response      &lt;dbl&gt; 9.40e-38, 1.04e-35, 3.03e-39, 2.17…
$ fdr_YPS606_WT_EtOH_response         &lt;dbl&gt; 3.26e-35, 1.54e-33, 2.07e-36, 4.57…
$ logFC_YPS606_msn24dd_EtOH_response  &lt;dbl&gt; 0.78481798, 0.60949852, 1.73591634…
$ pvalue_YPS606_msn24dd_EtOH_response &lt;dbl&gt; 3.430000e-06, 8.401730e-04, 4.6400…
$ fdr_YPS606_msn24dd_EtOH_response    &lt;dbl&gt; 7.420000e-06, 1.398507e-03, 2.8600…
$ logFC_WT.vs.msn24dd_EtOH_response   &lt;dbl&gt; -6.815179, -7.152329, -4.958086, -…
$ pvalue_WT.vs.msn24dd_EtOH_response  &lt;dbl&gt; 6.34e-32, 2.53e-30, 1.35e-27, 3.23…
$ fdr_WT.vs.msn24dd_EtOH_response     &lt;dbl&gt; 3.65e-28, 7.28e-27, 2.59e-24, 4.65…
$ logFC_WT.vs.msn24dd_unstressed      &lt;dbl&gt; -0.144061475, -0.365016862, -0.126…
$ pvalue_WT.vs.msn24dd_unstressed     &lt;dbl&gt; 0.350436027, 0.041423492, 0.425528…
$ fdr_WT.vs.msn24dd_unstressed        &lt;dbl&gt; 0.998531082, 0.998531082, 0.998531…
$ logFC_WT.vs.msn24dd_EtOH_absolute   &lt;dbl&gt; -6.959241, -7.517346, -5.084564, -…
$ pvalue_WT.vs.msn24dd_EtOH_absolute  &lt;dbl&gt; 8.55e-37, 2.04e-35, 3.06e-36, 7.93…
$ fdr_WT.vs.msn24dd_EtOH_absolute     &lt;dbl&gt; 1.64e-33, 1.96e-32, 3.52e-33, 2.85…</code></pre>
</div>
</div>
<p>Answer the following questions (type your answers below each question, blank code chunks are included that you can paste or type commands into):</p>
<ol type="1">
<li>How many genes are included in this study?</li>
</ol>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the length of the list of unique gene ids in the dataframe</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(DE_results_msn24_EtOH<span class="sc">$</span>gene_id))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 5756</code></pre>
</div>
</div>
<p>There are 5756 unique genes in the dataset.</p>
<ol start="2" type="1">
<li>Which gene has the highest log fold change in the <em>msn2/4</em>ΔΔ mutant EtOH response?</li>
</ol>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Reference the dataset</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>DE_results_msn24_EtOH <span class="sc">%&gt;%</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Sort the dataset by the log fold-change of the mutant ethanol response (descending = highest first)</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(logFC_YPS606_msn24dd_EtOH_response)) <span class="sc">%&gt;%</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get the first value in the dataset, which will have the highest value of the mutant ethanol response</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>(<span class="dv">1</span>) </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 18
  gene_id common_name annotation   logFC_YPS606_WT_etoh…¹ pvalue_YPS606_WT_eto…²
  &lt;chr&gt;   &lt;chr&gt;       &lt;chr&gt;                         &lt;dbl&gt;                  &lt;dbl&gt;
1 YJL052W TDH1        Glyceraldeh…                   10.1               1.34e-36
# ℹ abbreviated names: ¹​logFC_YPS606_WT_etoh_response,
#   ²​pvalue_YPS606_WT_etoh_response
# ℹ 13 more variables: fdr_YPS606_WT_EtOH_response &lt;dbl&gt;,
#   logFC_YPS606_msn24dd_EtOH_response &lt;dbl&gt;,
#   pvalue_YPS606_msn24dd_EtOH_response &lt;dbl&gt;,
#   fdr_YPS606_msn24dd_EtOH_response &lt;dbl&gt;,
#   logFC_WT.vs.msn24dd_EtOH_response &lt;dbl&gt;, …</code></pre>
</div>
</div>
<p>The TDH1 gene, which codes for glyceraldehyde-3-phosphate dehydrogenase isozyme, has the highest log fold change in the <em>msn2/4</em>ΔΔ mutant EtOH response- ~9.06.</p>
<ol start="3" type="1">
<li>How many heat shock protein (HSP) genes are differentially expressed (FDR &lt; 0.01) in unstressed conditions for the mutant?</li>
</ol>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Reference the dataset</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>DE_results_msn24_EtOH <span class="sc">%&gt;%</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get all rows where the FDR of the mutant unstressed condition is less than 0.01</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    fdr_WT.vs.msn24dd_unstressed <span class="sc">&lt;</span> <span class="fl">0.01</span>,</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get all rows where the common name includes the string "HSP"</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">grepl</span>(<span class="st">"HSP"</span>, common_name, <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 18
  gene_id common_name annotation   logFC_YPS606_WT_etoh…¹ pvalue_YPS606_WT_eto…²
  &lt;chr&gt;   &lt;chr&gt;       &lt;chr&gt;                         &lt;dbl&gt;                  &lt;dbl&gt;
1 YDR258C HSP78       Oligomeric …                   5.10               1.42e-29
2 YPL240C HSP82       Hsp90 chape…                   4.74               7.28e-28
# ℹ abbreviated names: ¹​logFC_YPS606_WT_etoh_response,
#   ²​pvalue_YPS606_WT_etoh_response
# ℹ 13 more variables: fdr_YPS606_WT_EtOH_response &lt;dbl&gt;,
#   logFC_YPS606_msn24dd_EtOH_response &lt;dbl&gt;,
#   pvalue_YPS606_msn24dd_EtOH_response &lt;dbl&gt;,
#   fdr_YPS606_msn24dd_EtOH_response &lt;dbl&gt;,
#   logFC_WT.vs.msn24dd_EtOH_response &lt;dbl&gt;, …</code></pre>
</div>
</div>
<p>There are two HSP proteins that are differentially expressed (FDR &lt; 0.01) in unstressed conditions for the mutant.</p>
<ol start="4" type="1">
<li>Do the genes with the largest magnitude fold changes <u>tend</u> to have the smallest <em>p</em>-values? Why might there be cases where there’s a mismatch between the magnitude of fold change and <em>p</em>-values?</li>
</ol>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>a1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> DE_results_msn24_EtOH, <span class="fu">aes</span>(<span class="at">x =</span> logFC_YPS606_msn24dd_EtOH_response,</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(pvalue_YPS606_msn24dd_EtOH_response))) <span class="sc">+</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"msn2/4∆∆ EtOH response"</span>, </span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"logFC"</span>, </span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"-log(p value)"</span>)</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>a2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> DE_results_msn24_EtOH, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">abs</span>(logFC_YPS606_msn24dd_EtOH_response),</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(pvalue_YPS606_msn24dd_EtOH_response))) <span class="sc">+</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>) <span class="sc">+</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Absolute msn2/4∆∆ EtOH response"</span>, </span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"|logFC|"</span>, </span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"-log(p value)"</span>)</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>b1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> DE_results_msn24_EtOH, <span class="fu">aes</span>(<span class="at">x =</span> logFC_YPS606_WT_etoh_response,</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(pvalue_YPS606_WT_etoh_response))) <span class="sc">+</span></span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"WT EtOH response"</span>, </span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"logFC"</span>, </span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"-log(p value)"</span>)</span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>b2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> DE_results_msn24_EtOH, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">abs</span>(logFC_YPS606_WT_etoh_response),</span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(pvalue_YPS606_WT_etoh_response))) <span class="sc">+</span></span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>) <span class="sc">+</span></span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Absolute WT EtOH response"</span>, </span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"|logFC|"</span>, </span>
<span id="cb12-31"><a href="#cb12-31" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"-log(p value)"</span>)</span>
<span id="cb12-32"><a href="#cb12-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-33"><a href="#cb12-33" aria-hidden="true" tabindex="-1"></a>c1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb12-34"><a href="#cb12-34" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> DE_results_msn24_EtOH, <span class="fu">aes</span>(<span class="at">x =</span> logFC_WT.vs.msn24dd_EtOH_response,</span>
<span id="cb12-35"><a href="#cb12-35" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(pvalue_WT.vs.msn24dd_EtOH_response))) <span class="sc">+</span></span>
<span id="cb12-36"><a href="#cb12-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb12-37"><a href="#cb12-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"WT vs msn2/4∆∆ EtOH response"</span>, </span>
<span id="cb12-38"><a href="#cb12-38" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"logFC"</span>, </span>
<span id="cb12-39"><a href="#cb12-39" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"-log(p value)"</span>)</span>
<span id="cb12-40"><a href="#cb12-40" aria-hidden="true" tabindex="-1"></a>c2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb12-41"><a href="#cb12-41" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> DE_results_msn24_EtOH, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">abs</span>(logFC_WT.vs.msn24dd_EtOH_response),</span>
<span id="cb12-42"><a href="#cb12-42" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(pvalue_WT.vs.msn24dd_EtOH_response))) <span class="sc">+</span></span>
<span id="cb12-43"><a href="#cb12-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb12-44"><a href="#cb12-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>) <span class="sc">+</span></span>
<span id="cb12-45"><a href="#cb12-45" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Absolute WT vs msn2/4∆∆ EtOH response"</span>, </span>
<span id="cb12-46"><a href="#cb12-46" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"|logFC|"</span>, </span>
<span id="cb12-47"><a href="#cb12-47" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"-log(p value)"</span>)</span>
<span id="cb12-48"><a href="#cb12-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-49"><a href="#cb12-49" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(b1, b2, a1, a2, c1, c2, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">nrow =</span> <span class="dv">3</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'
`geom_smooth()` using formula = 'y ~ x'
`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Homework_Assignment_1_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The genes with the largest magnitude fold change tend to have the smallest p-values (small p-value -&gt; large -log(p-value) because large changes in expression are typically significant. There might be a mismatch between the magnitude of fold change and the p-values</p>
<ol start="5" type="1">
<li>Which isoform of phosphoglucomutase is upregulated in response to ethanol stress, <em>PGM1</em> or <em>PGM2</em>? Do you think <em>msn2/4</em> is responsible for this difference? Why or why not?</li>
</ol>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Reference the dataset</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>DE_results_msn24_EtOH <span class="sc">%&gt;%</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get all the rows where the word phosphoglucomutase is present in the gene annotation </span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">grepl</span>(<span class="st">"phosphoglucomutase"</span>, annotation, <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Sort the dataframe by common name</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(common_name) <span class="sc">%&gt;%</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get only the relevant columns: common name and EtOH response</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(common_name, </span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>         logFC_YPS606_WT_etoh_response, </span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>         logFC_YPS606_msn24dd_EtOH_response, </span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>         logFC_WT.vs.msn24dd_EtOH_response)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 4
  common_name logFC_YPS606_WT_etoh_response logFC_YPS606_msn24dd_EtOH_response
  &lt;chr&gt;                               &lt;dbl&gt;                              &lt;dbl&gt;
1 PGM1                               -0.936                             -1.18 
2 PGM2                                7.60                               0.785
# ℹ 1 more variable: logFC_WT.vs.msn24dd_EtOH_response &lt;dbl&gt;</code></pre>
</div>
</div>
<p>PGM1 is mildly downregulated in both the WT ethanol response and the mutant ethanol response. PGM2 is highly upregulated in the WT ethanol response but only mildly upregulated in the mutant ethanol response. This indicates that the PGM2 isoform of phosphoglucomutase is normally upregulated during ethanol stress, a response that is not present to the same extent in the mutant. Based on this information, the <em>msn2/4</em>ΔΔ mutation is therefore associated with an impaired ethanol stress response and <em>could be</em> (but is not definitevly) responsible for the difference.</p>
<p><strong>Collaborators:</strong> n/a</p>
<p><strong>References:</strong> n/a</p>
<p><strong>AI Use:</strong> n/a</p>
<p>System information for reproducibility:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>pander<span class="sc">::</span><span class="fu">pander</span>(<span class="fu">sessionInfo</span>())</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
<div class="cell-output-display">
<p><strong>R version 4.5.2 (2025-10-31)</strong></p>
<p><strong>Platform:</strong> aarch64-apple-darwin20</p>
<p><strong>locale:</strong> en_US.UTF-8||en_US.UTF-8||en_US.UTF-8||C||en_US.UTF-8||en_US.UTF-8</p>
<p><strong>attached base packages:</strong> <em>stats</em>, <em>graphics</em>, <em>grDevices</em>, <em>utils</em>, <em>datasets</em>, <em>methods</em> and <em>base</em></p>
<p><strong>other attached packages:</strong> <em>cowplot(v.1.2.0)</em>, <em>ggplot2(v.4.0.1)</em>, <em>dplyr(v.1.1.4)</em> and <em>readr(v.2.1.6)</em></p>
<p><strong>loaded via a namespace (and not attached):</strong> <em>Matrix(v.1.7-4)</em>, <em>bit(v.4.6.0)</em>, <em>gtable(v.0.3.6)</em>, <em>jsonlite(v.2.0.0)</em>, <em>compiler(v.4.5.2)</em>, <em>crayon(v.1.5.3)</em>, <em>Rcpp(v.1.1.1)</em>, <em>tidyselect(v.1.2.1)</em>, <em>parallel(v.4.5.2)</em>, <em>splines(v.4.5.2)</em>, <em>scales(v.1.4.0)</em>, <em>yaml(v.2.3.12)</em>, <em>fastmap(v.1.2.0)</em>, <em>lattice(v.0.22-7)</em>, <em>R6(v.2.6.1)</em>, <em>labeling(v.0.4.3)</em>, <em>generics(v.0.1.4)</em>, <em>curl(v.7.0.0)</em>, <em>knitr(v.1.51)</em>, <em>tibble(v.3.3.1)</em>, <em>pander(v.0.6.6)</em>, <em>pillar(v.1.11.1)</em>, <em>RColorBrewer(v.1.1-3)</em>, <em>tzdb(v.0.5.0)</em>, <em>rlang(v.1.1.7)</em>, <em>utf8(v.1.2.6)</em>, <em>xfun(v.0.55)</em>, <em>S7(v.0.2.1)</em>, <em>bit64(v.4.6.0-1)</em>, <em>cli(v.3.6.5)</em>, <em>mgcv(v.1.9-3)</em>, <em>withr(v.3.0.2)</em>, <em>magrittr(v.2.0.4)</em>, <em>digest(v.0.6.39)</em>, <em>grid(v.4.5.2)</em>, <em>vroom(v.1.6.7)</em>, <em>rstudioapi(v.0.17.1)</em>, <em>hms(v.1.1.4)</em>, <em>nlme(v.3.1-168)</em>, <em>lifecycle(v.1.0.5)</em>, <em>vctrs(v.0.6.5)</em>, <em>evaluate(v.1.0.5)</em>, <em>glue(v.1.8.0)</em>, <em>farver(v.2.1.2)</em>, <em>rmarkdown(v.2.30)</em>, <em>tools(v.4.5.2)</em>, <em>pkgconfig(v.2.0.3)</em> and <em>htmltools(v.0.5.9)</em></p>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>